Added cryptography section

dwaite opened the pull request.

dwaite committed:
Added cryptography section

quartzjer commented on org/v2.md:
```diff
@@ -193,6 +193,88 @@ representation of an inner packet into the BODY of an outer one. As a
+A minimum 2048 bit key size is required for RSA keys. Implementations should
+refuse to communicate with a party using shorter keys.
+  
+The algorithms used are RSA with EME-OAEP (SHA-1, MFG1) for encrypting the 
```
I've never heard of "EME"? Would it be useful to also link to the OAEP spec here (and mention that it's 'padding', and pkcsv1.5 right?)?

dwaite replied:
RSA encryption does not have padding - it converts the bytes to a numeric value . Thats why the encrypted value cannot exceed the size of the RSA key.

dwaite replied:
Wikipedia links are a good idea. I'll add them.

quartzjer commented on org/v2.md:
```diff
@@ -193,6 +193,88 @@ representation of an inner packet into the BODY of an outer one. As a
+Top level packets do not have any addressing, instead relying on the ip/port 
+corresponding to a single hashname. You may be able to determine the sender of 
+an [`open`](#open) packet by attempting to verify the embedded RSA signature 
+against known public keys.
```
Is this worth trying to protect? I'm thinking it might be, to ward off later traffic analysis... We'd just need to aes encrypt the signature first, the secret being a combination of the public key and line id, w/ the same iv given.

dwaite replied:
Or possibly both sign and encrypt it. However, you could also track hashnames->ip/port without connecting to the hashnames in question, just other members in the network. Not quite as foolproof though

quartzjer replied:
I thought about it a lot, and I believe that it's a pretty big opening for network analysis (record all open packets at the isp level when looking for specific hashnames to appear) and needs to be protected from since it's very easy to do so right now.  Create the signature just like it is now, but then aes encrypt it, using the sha256 of the ecc key + line id as the aes key, and the attached iv... we'll have to discuss more in theroom or in person this week, but it's a great catch, thanks :)

quartzjer commented on org/v2.md:
```diff
@@ -193,6 +193,88 @@ representation of an inner packet into the BODY of an outer one. As a
+an [`open`](#open) packet by attempting to verify the embedded RSA signature 
+against known public keys.
+
+Use of a node name derived from a public key (vs. one chosen randomly) also 
```
Where is it good to point out that for validation, the attached DER format public keys need to be parsed/re-encoded before hashed, to ensure that the asn1 formatting is consistent?

dwaite replied:
Probably at the very beginning when talking about hashnames. Created a new section to talk about hashnames specifically.

quartzjer commented on org/v2.md:
```diff
@@ -193,6 +193,88 @@ representation of an inner packet into the BODY of an outer one. As a
+
+A public EC key pair is created when you send an [`open`](#open) packet. The 
+public key is exchanged encrypted with the recipient's RSA public key. Due to
+less implementations, EC keys are exchanged in "uncompressed" encoding.
```
I believe the official spec for that encoding is http://www.secg.org/collateral/sec1_final.pdf section 2.3.3 right? As I'm finding out, it's not that common of a format for libraries to produce/provide by default :/

dwaite replied:
Added elliptic curve info as well, but still needs justification of the curve choice.

dwaite committed:
Second revision.

- Added several additional links to wikipedia and to cryptography papers
- Re-stylized Markdown to use link references (so we aren't inlining tons
  of URLs)
- Edit document to use implicit link references where possible to further
  simplify/prettify plaintext version of document.
- Change some existing URLs to use HTTPS rather than HTTP

dwaite commented:
Only outstanding bit is justification of the choice for NIST P-256.

quartzjer commented:
NIST P-256 was chosen since it's one of the commonly well supported curves of decent size.  I expect we'll support more in the future (the chosen one could be easily specified inside the encrypted part of the open packet) but this looked like a good default to start with.
